src/shared/sam/cli/samCliConfiguration.ts

```diff
@@ -3,85 +3,151 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import * as vscode from 'vscode'
-import { SettingsConfiguration } from '../../settingsConfiguration'
+import { getLogger } from '../../logger'
+import { fromPackage, migrateSetting, SettingsConfiguration } from '../../settingsConfiguration'
+import { stripUndefined, toRecord } from '../../utilities/collectionUtils'
+import { once } from '../../utilities/functionUtils'
+import { ClassToInterfaceType, keys } from '../../utilities/tsUtils'
+import { DefaultSamCliLocationProvider, SamCliLocationProvider } from './samCliLocator'
 
-export interface SamCliConfiguration {
-    /** Gets the current SAM CLI location from the VSCode settings store. */
-    getSamCliLocation(): string | undefined
+// TODO(sijaden): remove after a few releases
+async function migrateLegacySettings() {
+    await migrateSetting(
+        {
+            key: 'aws.manuallySelectedBuckets',
+            type: SavedBuckets,
+        },
+        {
+            key: 'aws.samcli.manuallySelectedBuckets',
+            type: SavedBuckets,
+        }
+    )
 
-    /** Sets the SAM CLI location in the VSCode settings store. */
-    setSamCliLocation(location: string | undefined): Promise<void>
+    await migrateSetting(
+        {
+            key: 'aws.sam.enableCodeLenses',
+            type: Boolean,
+        },
+        {
+            key: 'aws.samcli.enableCodeLenses',
+            type: Boolean,
+        }
+    )
 
-    /**
-     * Initializes this SamCliConfiguration object from the VSCode user settings,
-     * or tries to auto-detect `sam` in the environment.
-     *
-     * @returns true if auto-detection failed; false if auto-detection succeeded or was not attempted.
-     */
-    initialize(): Promise<void>
+    await migrateSetting(
+        {
+            key: 'aws.samcli.lambda.timeout',
+            type: Number,
+        },
+        {
+            key: 'aws.samcli.lambdaTimeout',
```

Comment:
Naming a setting with `camelCase` causes the name to be rendered as "Lambda Timeout" instead of "lambda -> Timeout"

Dotted names describe _sections_